chitchat/src/failure_detector.rs

```diff
@@ -182,102 +185,107 @@ struct SamplingWindow {
-            let interval = last_value.elapsed();
+        let now = Instant::now();
+        if let Some(last_value) = self.last_heartbeat {
+            let interval = now.duration_since(last_value);
             if interval <= self.max_interval {
                 self.intervals.append(interval.as_secs_f64());
             }
         } else {
-            self.intervals.append(self.initial_interval.as_secs_f64());
-        };
-        self.last_heartbeat = Some(Instant::now());
+            // This is our first heartbeat.
+            // No way to compute an interval.
+            // This is fine.
+        }
+        self.last_heartbeat = Some(now);
     }
 
     /// Computes the sampling window's phi value.
-    pub fn phi(&self) -> f64 {
-        if let Some(last_heartbeat) = self.last_heartbeat {
-            assert!(self.intervals.mean() > 0.0);
-            let elapsed_time = last_heartbeat.elapsed().as_secs_f64();
-            elapsed_time / self.intervals.mean()
-        } else {
-            // if we phi is called before we have a sample, we assume the node isn't really alive.
-            f64::INFINITY
-        }
+    /// Returns `None` if have not received two heartbeat yet.
+    pub fn phi(&self) -> Option<f64> {
+        // We avoid computing phi if we have only received one heartbeat.
+        // It could be data from an old dead node after all.
+        let len_non_zero = NonZeroUsize::new(self.intervals.len())?;
+        let sum = self.intervals.sum();
+        let last_heartbeat = self.last_heartbeat?;
+        let interval_mean = self.additive_smoothing.compute_mean(len_non_zero, sum);
+        let elapsed_time = last_heartbeat.elapsed().as_secs_f64();
+        Some(elapsed_time / interval_mean)
     }
 }
 
 /// An array that retains a fixed number of streaming values.
 #[derive(Debug)]
 struct BoundedArrayStats {
     /// The values.
-    data: Vec<f64>,
-    /// Number of accumulated values.
-    size: usize,
+    values: Box<[f64]>,
     /// Is the values array filled?
     is_filled: bool,
-    /// Position of the index within the values array.
+    /// Position of the next value to be written in the values array.
     index: usize,
     /// The accumulated sum of values.
     sum: f64,
-    /// The accumulated mean of values.
-    mean: f64,
 }
 
 impl BoundedArrayStats {
-    pub fn new(size: usize) -> Self {
+    pub fn with_capacity(capacity: usize) -> Self {
         Self {
-            data: vec![0.0; size],
-            size,
+            values: vec![0.0; capacity].into_boxed_slice(),
             is_filled: false,
             index: 0,
             sum: 0.0,
-            mean: 0.0,
         }
     }
 
     /// Returns the mean.
-    pub fn mean(&self) -> f64 {
-        self.mean
+    pub fn sum(&self) -> f64 {
+        self.sum
     }
 
-    /// Appends a new value and updates the statistics.
-    pub fn append(&mut self, interval: f64) {
-        if self.index == self.size {
```

Comment:
I really dislike this kind of code because it is impossible to put a meaning to `self.index`.